kie-server-parent/kie-server-services/kie-server-services-task-assigning/kie-server-services-task-assigning-planning/src/main/java/org/kie/server/services/taskassigning/planning/util/UserUtil.java

```diff
@@ -50,15 +50,31 @@ public static boolean isUser(String userType) {
         return UserType.USER.equals(userType);
     }
 
-    public static List<Task> extractTasks(User user, Predicate<Task> predicate) {
-        List<Task> result = new ArrayList<>();
-        Task nextTask = user != null ? user.getNextTask() : null;
-        while (nextTask != null) {
-            if (predicate.test(nextTask)) {
-                result.add(nextTask);
-            }
-            nextTask = nextTask.getNextTask();
+    /**
+     * Given a Collection of external users, normally returned by the UserSystemService, produces the filtering of the
+     * elements with the following criteria:
+     * 1) Null elements are removed from the result.
+     * 2) Duplicate users are removed from the result. Since the external users implementation is unknown and the identifier
+     * of a user is the "id", two users are considered to be the same if they have the same "id". Meaning that if two users
+     * with the same "id" are encountered only one instance will be included in the result, typically the first instance
+     * found.
+     * @param externalUsers a collection with the external users to filter.
+     * @return a Stream with the filtered users.
+     */
+    public static Stream<org.kie.server.services.taskassigning.user.system.api.User> filterDuplicates(Collection<org.kie.server.services.taskassigning.user.system.api.User> externalUsers) {
+        if (externalUsers == null) {
+            return Stream.empty();
+        } else {
+            final Set<String> included = new HashSet<>();
+            return externalUsers.stream()
+                    .filter(Objects::nonNull)
+                    .filter(externalUser -> {
```

Comment:
Java streams offer a method `distinct` [1] to avoid duplicate objects in the stream.  The only prerequisite here is that the `User` class has the `equals` method based on the `id`.

[1] https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#distinct--

Reply:
Yes, but note that the User system instance we are processing here is an "unknown" implementation that comes from the particular configured UserSystemItegration. So at this point we can't rely that such equals implementation based on the userId will exist. This explains current implementation.